sdk_extensions/zpages/src/test/java/io/opentelemetry/sdk/extensions/zpages/SpanBucketTest.java

```diff
@@ -0,0 +1,111 @@
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Unit tests for {@link SpanBucket}. */
+@RunWith(JUnit4.class)
+public final class SpanBucketTest {
+  private static final String SPAN_NAME = "span";
+  private static final int LATENCY_BUCKET_SIZE = 16;
+  private static final int ERROR_BUCKET_SIZE = 8;
+  private final TracerSdkProvider tracerSdkProvider = TracerSdkProvider.builder().build();
+  private final Tracer tracer = tracerSdkProvider.get("SpanBucketTest");
+
+  @Test
+  public void verifyLatencyBucketSizeLimit() {
+    SpanBucket latencyBucket = new SpanBucket(true);
+    Span[] spans = new Span[LATENCY_BUCKET_SIZE + 1];
+    for (int i = 0; i < LATENCY_BUCKET_SIZE + 1; i++) {
+      spans[i] = tracer.spanBuilder(SPAN_NAME).startSpan();
+      latencyBucket.add((ReadableSpan) spans[i]);
+      spans[i].end();
+    }
+    List<ReadableSpan> bucketSpans = new ArrayList<>();
+    latencyBucket.addTo(bucketSpans);
+    /* The latency SpanBucket should have the most recent LATENCY_BUCKET_SIZE spans */
+    assertThat(latencyBucket.size()).isEqualTo(LATENCY_BUCKET_SIZE);
+    assertThat(bucketSpans.size()).isEqualTo(LATENCY_BUCKET_SIZE);
+    assertThat(bucketSpans).doesNotContain(spans[0]);
+    for (int i = 1; i < LATENCY_BUCKET_SIZE + 1; i++) {
+      assertThat(bucketSpans).contains(spans[i]);
+    }
+  }
+
+  @Test
+  public void verifyErrorBucketSizeLimit() {
+    SpanBucket errorBucket = new SpanBucket(false);
+    Span[] spans = new Span[ERROR_BUCKET_SIZE + 1];
+    for (int i = 0; i < ERROR_BUCKET_SIZE + 1; i++) {
+      spans[i] = tracer.spanBuilder(SPAN_NAME).startSpan();
+      errorBucket.add((ReadableSpan) spans[i]);
+      spans[i].end();
+    }
+    List<ReadableSpan> bucketSpans = new ArrayList<>();
+    errorBucket.addTo(bucketSpans);
+    /* The error SpanBucket should have the most recent ERROR_BUCKET_SIZE spans */
+    assertThat(errorBucket.size()).isEqualTo(ERROR_BUCKET_SIZE);
+    assertThat(bucketSpans.size()).isEqualTo(ERROR_BUCKET_SIZE);
+    assertThat(bucketSpans).doesNotContain(spans[0]);
+    for (int i = 1; i < ERROR_BUCKET_SIZE + 1; i++) {
+      assertThat(bucketSpans).contains(spans[i]);
+    }
+  }
+
+  @Test
+  public void verifyThreadSafety() throws InterruptedException {
+    int numberOfThreads = 4;
+    int numberOfSpans = 4;
+    SpanBucket spanBucket = new SpanBucket(true);
+    final CountDownLatch startSignal = new CountDownLatch(1);
+    final CountDownLatch endSignal = new CountDownLatch(numberOfThreads);
+    for (int i = 0; i < numberOfThreads; i++) {
+      new Thread(
+              () -> {
+                try {
+                  startSignal.await();
+                  for (int j = 0; j < numberOfSpans; j++) {
+                    Span span = tracer.spanBuilder(SPAN_NAME).startSpan();
+                    spanBucket.add((ReadableSpan) span);
+                    span.end();
+                  }
+                  endSignal.countDown();
```

Comment:
because we're relying on this, we should put a timeout on the test annotation, just in case things go badly awry. That way, we won't hang indefinitely waiting for things to finish.